wallets/dcr/syncnotification.go

```diff
@@ -0,0 +1,678 @@
+package dcr
+
+import (
+	"math"
+	"time"
+
+	"github.com/planetdecred/dcrlibwallet/spv"
+	"golang.org/x/sync/errgroup"
+)
+
+func (w *Wallet) spvSyncNotificationCallbacks() *spv.Notifications {
+	return &spv.Notifications{
+		PeerConnected: func(peerCount int32, addr string) {
+			w.handlePeerCountUpdate(peerCount)
+		},
+		PeerDisconnected: func(peerCount int32, addr string) {
+			w.handlePeerCountUpdate(peerCount)
+		},
+		Synced:                       w.synced,
+		FetchHeadersStarted:          w.fetchHeadersStarted,
+		FetchHeadersProgress:         w.fetchHeadersProgress,
+		FetchHeadersFinished:         w.fetchHeadersFinished,
+		FetchMissingCFiltersStarted:  w.fetchCFiltersStarted,
+		FetchMissingCFiltersProgress: w.fetchCFiltersProgress,
+		FetchMissingCFiltersFinished: w.fetchCFiltersEnded,
+		DiscoverAddressesStarted:     w.discoverAddressesStarted,
+		DiscoverAddressesFinished:    w.discoverAddressesFinished,
+		RescanStarted:                w.rescanStarted,
+		RescanProgress:               w.rescanProgress,
+		RescanFinished:               w.rescanFinished,
+	}
+}
+
+func (w *Wallet) handlePeerCountUpdate(peerCount int32) {
+	w.syncData.mu.Lock()
+	w.syncData.connectedPeers = peerCount
+	shouldLog := w.syncData.showLogs && w.syncData.syncing
+	w.syncData.mu.Unlock()
+
+	for _, syncProgressListener := range w.syncProgressListeners() {
+		syncProgressListener.OnPeerConnectedOrDisconnected(peerCount)
+	}
+
+	if shouldLog {
+		if peerCount == 1 {
+			log.Infof("Connected to %d peer on %s.", peerCount, w.chainParams.Name)
+		} else {
+			log.Infof("Connected to %d peers on %s.", peerCount, w.chainParams.Name)
+		}
+	}
+}
+
+// Fetch CFilters Callbacks
+
+func (w *Wallet) fetchCFiltersStarted(walletID int) {
+	w.syncData.mu.Lock()
+	w.syncData.activeSyncData.syncStage = CFiltersFetchSyncStage
+	w.syncData.activeSyncData.cfiltersFetchProgress.beginFetchCFiltersTimeStamp = time.Now().Unix()
+	w.syncData.activeSyncData.cfiltersFetchProgress.totalFetchedCFiltersCount = 0
+	showLogs := w.syncData.showLogs
+	w.syncData.mu.Unlock()
+
+	if showLogs {
+		log.Infof("Step 1 of 3 - fetching %d block headers.")
+	}
+}
+
+func (w *Wallet) fetchCFiltersProgress(walletID int, startCFiltersHeight, endCFiltersHeight int32) {
+
+	// lock the mutex before reading and writing to w.syncData.*
+	w.syncData.mu.Lock()
+
+	if w.syncData.activeSyncData.cfiltersFetchProgress.startCFiltersHeight == -1 {
+		w.syncData.activeSyncData.cfiltersFetchProgress.startCFiltersHeight = startCFiltersHeight
+	}
+
+	// wallet := w.WalletWithID(walletID)
+	w.syncData.activeSyncData.cfiltersFetchProgress.totalFetchedCFiltersCount += endCFiltersHeight - startCFiltersHeight
+
+	totalCFiltersToFetch := w.getBestBlock() - w.syncData.activeSyncData.cfiltersFetchProgress.startCFiltersHeight
+	// cfiltersLeftToFetch := totalCFiltersToFetch - w.syncData.activeSyncData.cfiltersFetchProgress.totalFetchedCFiltersCount
```

Comment:
I presume this commented code is useless, why not delete it?

Reply:
thanks for pointing out, should be uncommented